ouroboros-consensus/src/unstable-consensus-testlib/Test/Ouroboros/Consensus/DiffusionPipelining.hs

```diff
@@ -0,0 +1,45 @@
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications    #-}
+
+module Test.Ouroboros.Consensus.DiffusionPipelining (prop_diffusionPipeliningSubsequenceConsistency) where
+
+import           Control.Exception (assert)
+import           Data.Either (isRight)
+import           Data.Proxy
+import           Ouroboros.Consensus.Block
+import           Test.QuickCheck
+
+-- | See /Consistent validity under subsequences/ in
+-- 'BlockSupportsDiffusionPipelining'.
+prop_diffusionPipeliningSubsequenceConsistency ::
+     forall blk.
+     BlockSupportsDiffusionPipelining blk
+  => Proxy blk
+  -> [TentativeHeaderView blk]
+     -- ^ Have to satisfy the pipelining criterion.
+  -> Property
+prop_diffusionPipeliningSubsequenceConsistency _ thvs =
+    assert (isRight $ satisfyPipeliningCriterion thvs) $
```

Comment:
The input list should satisfy the pipelining criterion as a precondition, as the property doesn't make sense otherwise (ie, it is fine if the property fails when `isLeft $ satisfyPipeliningCriterion thvs`). So the assertion is intended to signal the difference between testing "A and B" and "A implies B" (in both cases, you only have to test something when A is true, but A being false only invalidates the first statement, not the second one). But that is minor, discarding such cases or making it part of the property also seem fine.